wpilibc/src/main/native/include/frc/simulation/LinearSystemSim.h

```diff
@@ -0,0 +1,124 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2020 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <array>
+
+#include <Eigen/Core>
+#include <units/current.h>
+#include <units/time.h>
+
+#include "frc/StateSpaceUtil.h"
+#include "frc/system/LinearSystem.h"
+
+namespace frc {
+namespace sim {
+/**
+ * Represents a simulated generic linear system.
+ */
+template <int States, int Inputs, int Outputs>
+class LinearSystemSim {
+ public:
+  /**
+   * Creates a simulated generic linear system.
+   *
+   * @param system             The system to simulate.
+   * @param addNoise           Whether the sim should automatically add some
+   *                           measurement noise.
+   * @param measurementStdDevs The standard deviations of the measurement noise.
+   */
+  LinearSystemSim(const LinearSystem<States, Inputs, Outputs>& system,
+                  bool addNoise = false,
+                  const std::array<double, Outputs>& measurementStdDevs = {0.0})
+      : m_plant(system),
+        m_shouldAddNoise(addNoise),
+        m_measurementStdDevs(measurementStdDevs) {
+    m_x = Eigen::Matrix<double, States, 1>::Zero();
+    m_y = Eigen::Matrix<double, Outputs, 1>::Zero();
+    m_u = Eigen::Matrix<double, Inputs, 1>::Zero();
+  }
+
+  /**
+   * Returns whether the sim should add noise to the measurements.
+   * @return Whether the sim should add noise to the measurements.
+   */
+  bool ShouldAddNoise() const { return m_shouldAddNoise; }
+
+  /**
+   * Returns the current output of the plant.
+   * @return The current output of the plant.
+   */
+  const Eigen::Matrix<double, Outputs, 1>& Y() const { return m_y; }
```

Comment:
Also add `double Y(int i) const { return m_y(i); }`. This will be useful in derived classes.